core-bundle/src/Routing/UrlGenerator.php

```diff
@@ -215,4 +222,57 @@ private function getAutoItems(array $parameters): array
 
         return [];
     }
+
+    private function getRoute(string $name, array &$parameters)
+    {
+        $languagePrefix = '';
+        $requirements = [];
+
+        if ($this->legacyRouting) {
+            $urlSuffix = $this->urlSuffix;
+
+            if ($this->prependLocale) {
+                $languagePrefix = '/{_locale}';
+                $requirements['_locale'] = '[a-z]{2}(\-[A-Z]{2})?';
+            } else {
+                unset($parameters['_locale']);
+            }
+        } else {
+            $rootPage = $this->findRootPage($parameters);
+            $urlSuffix = $rootPage->urlSuffix;
+
+            if ($rootPage->languagePrefix) {
+                $languagePrefix = '/'.$rootPage->languagePrefix;
+            }
+
+            unset($parameters['_locale']);
+        }
+
+        if ('index' === $name) {
+            return new Route($languagePrefix.'/', [], $requirements);
+        }
+
+        $requirements['alias'] = '.+';
+
+        return new Route($languagePrefix.'/{alias}'.$urlSuffix, [], $requirements);
+    }
+
+    private function findRootPage(array $parameters): PageModel
+    {
+        /** @var PageModel $pageAdapter */
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+
+        $rootPage = $pageAdapter->findFirstPublishedRootByHostAndLanguage(
+            $parameters['_domain'] ?? '',
+            $parameters['_locale'] ?? ''
+        );
+
+        if (null === $rootPage) {
+            throw new \RuntimeException('The Contao\CoreBundle\Routing\UrlGenerator requires a domain and locale or legacy routing. Configure "prepend_locale" or "url_suffix" in the Contao bundle.');
```

Comment:
Random nitpick: I'd use `"` and the `::class` notation just to ensure the error message stays correct, even if you decide to move the class :)

Reply:
Changed in f384d55c14f51bf148889db75552b25c88f07fab